### PR TITLE
Require private configs to be fully qualified when set on command-line

### LINKED ISSUE #14622: Should setting private configs on the command line require the module name?
Currently, private configs can be set on the command line without specifying the module name, but I wonder if it would make more sense to always require the module name for private configs. e.g. for:

```chpl
module M1 {
  private config const v = false;
}

module M2 {
  use M1;
  config const v = false;
  proc main() { writeln(v); }
}
```

```
$ ./test --v
<command-line arg>:1: error: Configuration variable 'v' is defined in more than one module.  Use '--help' for a list of configuration variables and '-s<module>.v' to disambiguate.
```

Should non qualified configs setting on the command line only work for public configs, and the above test should work, setting `v` in M2?

### PR DESCRIPTION
**Pull request by bradcray**
This PR changes the command-line parsing of `private config`s such that they must always be fully qualified when specified on the command-line, as proposed in issue #14622 

The approach taken for configs on the executable command-line is fairly simple: When the generated code registers the configs, it now passes a new 4th argument which specifies whether or not the config is private (implemented in codegen.cpp).  The runtime stores this flag in the table of configs and uses it to ensure that only public configs are set in an unqualfied manner and to flag configs that are private as such in the executable's `--help` output (implemented in config.c / config.h).

The approach taken for compile-time configs is slightly more complicated due to the fact that in the existing code structure, when we were processing the configs, we didn't know whether or not they were private due to the fact that the parser hadn't yet processed the `private` keyword when the configs are being set up.  Here I've taken the workable, but slightly lame, solution of storing a boolean flag (`parsingPrivate`) indicating whether or not we've processed a `private` keyword, checking it when we process the config, and then restoring it after the config has been processed.  The reason I call this "slightly lame" is that a general solution would work for complicated nested declarations like a private class that contains a public method that contains a private mumble... but my attempts to have the `public` cases similarly set and restore the variable ran into shift/reduce problems due to the fact that the `public` keyword is optional, so we're effectively trying to take actions before we can be sure which of the three cases (implicit, public, or private) we're taking.  Since I'm only using the flag for configs, which by nature can't have nested public/private declarations, this isn't actually a problem, but it means that if others tried to lean on this global more generally to determine whether the current declaration was public or private, they would get incorrect results.  All of the above is implemented in `chapel.ypp`.

Then, `build.cpp` was updated to look up all configs in a fully qualified manner and then only consider looking them up in an unqualified manner if the config is public.  If a config is set in both a fully and qualified manner, we now issue an error message saying that it's ambiguous which we should use, where before we generated a more confusing error message.  One might argue that we should just take the last flag on the command line rather than complaining, but our current config machinery doesn't store the order in which the flags were specified, and that effort is unrelated to the main goal of this PR, so I commented on it and left it for another day.

I also added two new tests:

* test/compflags/configs/privateConfigParam.chpl: which declares a public and a private config param and then tests setting them in various ways
* test/execflags/configs/privateconfigs.chpl: which tests various combinations of public and private configs

In terms of documentation, I added a paragraph to `usingchapel/executing.rst` describing the ability to use fully-qualified config names and the need to do so for private configs.  And I did a bit of wordsmithing in the spec in the configs section while I was reading it to see whether there was anything requiring updates there.

Updated existing tests of setting private configs to use fully-qualified names.

Resolves #14622.

TODO:
- [x] extend to LLVM back-end
- [x] extend to compile-time configs?
- [x] commit parser-generated files once review is done

### CODE REVIEW

**Comment by bradcray** [comment]
@lydia-duncan: Would you be able to review this PR?

File: test/compflags/configs/privateConfigParam.compopts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+-sfoo=true                 # privateConfigParam-ft.good
+-sM.foo=true               # privateConfigParam-tf.good
+-sM2.foo=true              # privateConfigParam-ft.good
+-sfoo=true -sM.foo=true    # privateConfigParam-tt.good
+-sfoo=true -sM2.foo=true   # privateConfigParam-ambig.good
+-sM.foo=true -sM.foo=false
+-sM.foo=false -sM.foo=true # privateConfigParam-tf.good
```
Comment (lydia-duncan) [inline test/compflags/configs/privateConfigParam.compopts]: Unimportant, but missing a newline here

File: test/execflags/configs/privateconfigs.execopts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+--aaa=0 --bbb=0 --ccc=0 --ddd=0
+--eee=0  # privateconfigs-eee.good
+--fff=0  # privateconfigs-fff.good
+--ggg=0  # privateconfigs-ggg.good
+--help   # privateconfigs-help.good
```
Comment (lydia-duncan) [inline test/execflags/configs/privateconfigs.execopts]: Ditto here

File: runtime/src/config.c [hunk 3]
```diff
@@ -212,8 +213,18 @@ void printConfigVarTable(void) {
...
+        if (configVar->setValue || configVar->private) {
+          fprintf(stdout, " (");
+          if (configVar->private) {
+            fprintf(stdout, "private");
+            if (configVar->setValue) {
+              fprintf(stdout, ", ");
+            }
+          }
+          if (configVar->setValue) {
+            fprintf(stdout, "configured to %s", configVar->setValue);
```
Comment (lydia-duncan) [inline runtime/src/config.c]: I expected to see a test updated/added in response to this change, but I don't see one.

Comment (bradcray) [reply runtime/src/config.c]: Assuming you're talking about the labeling of configs as private in the help output, the last case (`--help` / `privateconfigs-help.good`) in `test/execflags/configs/privateconfigs.chpl` is meant to cover this case.

Comment (lydia-duncan) [reply runtime/src/config.c]: I was expecting to see something with "(private, configured to " in it, based on how this code looks, but I don't see anything like that

Comment (bradcray) [reply runtime/src/config.c]: Oh that.  That's old code that didn't really "change" so much as moved (line 216 prior to this effort became line 225).  The reason for the refactoring was just to use the same parens for either message being printed out.  There are a bunch of tests in test/execflags/shannon/configs/help/ that ensure it's still working.

Comment (bradcray) [reply runtime/src/config.c]: Ah, but now I understand that you're saying you'd like to see a test that shows both messages.  That seems appropriate, let me add one.

**Pull request merged**